### PR TITLE
[quant] Add a quantized batch_norm operator

### PR DESCRIPTION
**Pull request by supriyar**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33109 [quant] Add Quantized BatchNorm2d module
* **#33080 [quant] Add a quantized batch_norm operator**

Summary:
Quantized batch norm for cases where batch norm cannot be fused with conv.
AVX2 implementation is from Caffe2.

Test Plan:
python test/test_quantized.py TestQuantizedOps.test_batch_norm

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D19861927](https://our.internmc.facebook.com/intern/diff/D19861927)

### CODE REVIEW

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -476,6 +476,11 @@
 
 - func: batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled) -> Tensor
 
+- func: quantized_batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor mean, Tensor var, float eps, float output_scale, int output_zero_point) -> Tensor
```
Comment (raghuramank100) [inline aten/src/ATen/native/native_functions.yaml]: We should also register the variant with the relu (quantized_batch_norm_relu), like other cases:https://github.com/pytorch/pytorch/blob/ecbf6f99e6a4e373105133b31534c9fb50f2acca/aten/src/ATen/native/quantized/cpu/qconv.cpp#L640

File: aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    float output_scale,
+    float* alpha_data,
+    float* beta_data) {
+  // Batch Normalization
+  // output(n, c, h, w)
+  //     = (input(n, c, h, w) - mean(c)) / sqrt(var(c) + eps) * weight(c)
+  //         + bias(c)
+  // We factor out inv_var(c) = 1 / sqrt(var(c) + eps).
+  for (int64_t c = 0; c < channels; c++) {
+    float inv_var = 1 / std::sqrt(var_data[c] + static_cast<float>(eps));
```
Comment (raghuramank100) [inline aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp]: Lets call in inv_sigma as this is the standard deviation, not the variance.

File: test/test_quantized.py [hunk 3]
```diff
@@ -1071,6 +1071,32 @@ def equal_ref(qX, qX2):
         self.assertEqual(qX.equal(qX), equal_ref(qX, qX))
         self.assertEqual(qX.equal(qX2), equal_ref(qX, qX2))
 
+    @given(X=hu.tensor(shapes=hu.array_shapes(min_dims=4, max_dims=4,
```
Comment (raghuramank100) [inline test/test_quantized.py]: Can we also test for qx being of type qint8?

**Review by raghuramank100 — CHANGES_REQUESTED** [review]
Looks good! A few requests:

File: aten/src/ATen/native/native_functions.yaml (hunk 1 above)
Comment (supriyar) [reply aten/src/ATen/native/native_functions.yaml]: Let's leave it at quantized batch_norm support for now and add relu fusion later (if required)

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 4]
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
...
+    int N,
+    int C,
+    int HxW,
+    const int in_zero_point,
+    const int out_zero_point,
+    const uint8_t* X,
+    const float* alpha,
+    const float* beta,
+    uint8_t* Y) {
+#if defined(__AVX2__)
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]:
```
#if defined(__AVX2__) && defined(__FMA__)
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 5]
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
   });
 }
 
+template <bool ReluFused>
+void q_batch_norm_kernel(
+    int N,
+    int C,
+    int HxW,
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: We sure we want to use `int` for all these shape parameters? It's probably better if this is at least fixed-width, and preferrably `int64_t` to match the rest of the index handling in pt

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 6]
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
...
+    const int in_zero_point,
+    const int out_zero_point,
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: zero points should definitely be `int64_t`, right?

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 7]
```diff
@@ -955,6 +961,126 @@ void qtopk_kernel(Tensor& values,
...
+  const __m256i out_zero_point_v = _mm256_set1_epi32(out_zero_point);
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+  for (int i = 0; i < outer_size; ++i) {
+    int n = C / kVLen * kVLen;
+    int r = C % kVLen;
+    const uint8_t* X_ptr = X + i * C;
+    uint8_t* Y_ptr = Y + i * C;
+    for (int j = 0; j < n; j += kVLen) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Can this instead be done 4x interleaved to ensure better pipeline occupancy? You can then replace most of these instructions with calls to Vec256<quint8>::quantize (https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/cpu/vec256/vec256_qint.h#L436)

That also cleans up the unnecessary shuffle instructions etc

File: aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp (hunk 2 above)
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp]: 1.0f

File: test/test_quantized.py (hunk 3 above)
Comment (supriyar) [reply test/test_quantized.py]: Added an assertion for type other than uint8. Updated test to use both qengines.

I'm not sure how we can make sure to test the AVX2 and default path as it is guarded by compile time macro. I am relying on CI to run on machines without AVX2.

File: aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp [hunk 8]
```diff
@@ -0,0 +1,167 @@
...
+  if (qx.numel() == 0) {
+    auto out = qx.clone();
+    return out;
+  }
+  int ndim = qx.dim();
+  TORCH_CHECK(ndim == 4, "Expecting the input tensor of rank 4.");
+  const int N = qx.size(0);
+  const int C = qx.size(1);
+  const int H = qx.size(2);
+  const int W = qx.size(3);
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp]: All of these are narrowing conversions `int` -> `int64_t`

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 6 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Yeah, updated them

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp (hunk 7 above)
Comment (jamesr66a) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: NB: I haven't compiled or run this so there might be some details off! :p

```
using Vec = Vec256<quint8>;

// Hoisted variables
auto in_zp_vec = Vec256<float>(static_cast<float>(in_zero_point));
auto fake_scale = Vec256<float>(1.0f);
auto scale_neg_zp_premul = fake_input_scale + in_zp.vec.neg();
auto out_zero_point_v = Vec(c10::quint8(out_zero_point));

for (int64_t j = 0; j < n; j += Vec::float_num_vecs()) {
  auto vals_q = Vec::loadu(X_ptr + j);
  // Fake scale of 1.0 here, should not affect performance (FMA in place of sub)
  auto vals_dq = vals_q.dequantize(fake_in_scale, in_zp_vec, scale_neg_zp_premul);
  for (int idx = 0; idx < vals_dq.size(); ++idx) {
    auto alpha_v = Vec256<float>(alpha + j * Vec::float_num_vecs() + idx);
    auto beta_v = Vec256<float>(beta + j * Vec::float_num_vecs() + idx);
    vals_dq[idx] = vec256::fmadd(vals_dq[idx], alpha_v, beta_v);
  }
  // Fake scale again
  auto outputs_q = Vec::quantize(vals_dq, /*output_scale=*/1.0f, out_zero_point, /*inv_output_scale=*/1.0f);
  if (ReluFused) {
    outputs_q = outputs_q.relu(out_zero_point_v);
  }
  outputs_q.store(Y_ptr + j);
}
```

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 9]
```diff
@@ -955,6 +961,89 @@ void qtopk_kernel(Tensor& values,
...
+    const uint8_t* X,
+    const float* alpha,
+    const float* beta,
+    uint8_t* Y) {
+#if defined(__AVX2__) && defined(__FMA__)
+
+  constexpr int kVLen = 8;
+  const int outer_size = N * HxW;
+  using Vec = Vec256<quint8>;
+  for (int i = 0; i < outer_size; ++i) {
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: I think we might be able to simplify this further, since--correct me if i'm wrong--this is essentially just an elementwise access pattern where we broadcast `alpha` and `beta` to all [N, H, W]. It might be pretty simple to just use TensorIterator instead of handling the loop nest explicitly. I can elaborate a bit more later

Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: Blocked by #33166

File: aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp [hunk 10]
```diff
@@ -955,6 +961,70 @@ void qtopk_kernel(Tensor& values,
...
+    const int64_t in_zero_point,
+    const int64_t out_zero_point,
+    const uint8_t* X,
+    const float* alpha,
+    const float* beta,
+    uint8_t* Y) {
+
+  constexpr int kVLen = 8;
+  const int64_t outer_size = N * HxW;
+  using Vec = Vec256<quint8>;
```
Comment (jamesr66a) [inline aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp]: We can generalize this with AT_FORALL_QINT_TYPES now that we've gotten rid of the bare intrinsics

**Review by jamesr66a — APPROVED** [review]
LG, just one comment
